src/librbd/Lock.cc

```diff
@@ -0,0 +1,620 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/Lock.h"
+#include "librbd/managed_lock/Policy.h"
+#include "librbd/managed_lock/AcquireRequest.h"
+#include "librbd/managed_lock/ReleaseRequest.h"
+#include "librbd/managed_lock/ReacquireRequest.h"
+#include "librbd/managed_lock/NoReleasePolicy.h"
+#include "cls/lock/cls_lock_client.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/WorkQueue.h"
+#include "librbd/Utils.h"
+#include <sstream>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Lock: "
+
+namespace librbd {
+
+using namespace managed_lock;
+using std::string;
+using util::detail::C_AsyncCallback;
+
+namespace {
+
+template <typename R>
+struct C_SendRequest : public Context {
+  R* request;
+  explicit C_SendRequest(R* request) : request(request) {
+  }
+  virtual void finish(int r) override {
+    request->send();
+  }
+};
+
+class ThreadPoolSingleton : public ThreadPool {
+public:
+  explicit ThreadPoolSingleton(CephContext *cct)
+    : ThreadPool(cct, "librbd::lock::thread_pool", "tp_librbd_lock", 1) {
+    start();
+  }
+  virtual ~ThreadPoolSingleton() {
+    stop();
+  }
+};
+
+} // anonymous namespace
+
+template <typename L>
+Lock<L>::Lock(librados::IoCtx &ioctx, const string& oid,
+           Policy *policy)
+  : m_ioctx(ioctx), m_cct(reinterpret_cast<CephContext *>(ioctx.cct())),
+    m_oid(oid),
+    m_policy(policy),
+    m_lock(util::unique_lock_name("librbd::Lock<L>::m_lock", this)),
+    m_state(STATE_UNLOCKED) {
+
+  if (m_policy == nullptr) {
```

Comment:
... and coming back to this now that I've seen how it's used, I think the policy can be scrapped at this level and be moved to back to ```ImageWatcher``` / ```ExclusiveLock```-only.